substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java

```diff
@@ -0,0 +1,290 @@
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+import org.graalvm.compiler.code.CompilationResult;
+import org.graalvm.compiler.code.SourceMapping;
+import org.graalvm.compiler.debug.DebugContext;
+import org.graalvm.compiler.graph.NodeSourcePosition;
+import org.graalvm.nativeimage.ImageSingletons;
+
+import com.oracle.graal.pointsto.infrastructure.OriginalClassProvider;
+import com.oracle.graal.pointsto.meta.AnalysisType;
+import com.oracle.objectfile.debuginfo.DebugInfoProvider;
+import com.oracle.svm.hosted.image.sources.SourceManager;
+import com.oracle.svm.hosted.meta.HostedMethod;
+import com.oracle.svm.hosted.meta.HostedType;
+
+import jdk.vm.ci.code.site.Mark;
+import jdk.vm.ci.meta.LineNumberTable;
+import jdk.vm.ci.meta.ResolvedJavaMethod;
+import jdk.vm.ci.meta.ResolvedJavaType;
+
+/**
+ * Implementation of the DebugInfoProvider API interface that allows type, code and heap data info
+ * to be passed to an ObjectFile when generation of debug info is enabled.
+ */
+class NativeImageDebugInfoProvider implements DebugInfoProvider {
+    private final DebugContext debugContext;
+    private final NativeImageCodeCache codeCache;
+    @SuppressWarnings("unused") private final NativeImageHeap heap;
+
+    NativeImageDebugInfoProvider(DebugContext debugContext, NativeImageCodeCache codeCache, NativeImageHeap heap) {
+        super();
+        this.debugContext = debugContext;
+        this.codeCache = codeCache;
+        this.heap = heap;
+    }
+
+    @Override
+    public Stream<DebugTypeInfo> typeInfoProvider() {
+        return Stream.empty();
+    }
+
+    @Override
+    public Stream<DebugCodeInfo> codeInfoProvider() {
+        return codeCache.compilations.entrySet().stream().map(entry -> new NativeImageDebugCodeInfo(entry.getKey(), entry.getValue()));
+    }
+
+    @Override
+    public Stream<DebugDataInfo> dataInfoProvider() {
+        return Stream.empty();
+    }
+
+    /**
+     * implementation of the DebugCodeInfo API interface that allows code info to be passed to an
+     * ObjectFile when generation of debug info is enabled.
+     */
+    private class NativeImageDebugCodeInfo implements DebugCodeInfo {
+        private final HostedMethod method;
+        private final ResolvedJavaType javaType;
+        private final CompilationResult compilation;
+        private Path fullFilePath;
+
+        NativeImageDebugCodeInfo(HostedMethod method, CompilationResult compilation) {
+            this.method = method;
+            HostedType declaringClass = method.getDeclaringClass();
+            Class<?> clazz = declaringClass.getJavaClass();
+            this.javaType = declaringClass.getWrapped();
+            this.compilation = compilation;
+            fullFilePath = ImageSingletons.lookup(SourceManager.class).findAndCacheSource(javaType, clazz);
+        }
+
+        @SuppressWarnings("try")
+        @Override
+        public void debugContext(Consumer<DebugContext> action) {
```

Comment:
This allow you to perfom `action` scoped within `scope("DebugCodeInfo", method))`

Reply:
The passed method allow you to filter the logging output. E.g.
```
mx native-image -jar svmbuild/vm/lib/graalvm/svm-driver.jar -H:GenerateDebugInfo=1 -H:Log=DebugCodeInfo,~DebugCodeInfo.Subranges -H:MethodFilter=JavaConstant.toString
```
gives
```
...
[native-image:29453]     (inline):   2,250.32 ms,  6.93 GB
[native-image:29453]    (compile):  15,444.89 ms,  6.93 GB
[native-image:29453]      compile:  21,605.21 ms,  6.93 GB
[Use -Dgraal.LogFile=<path> to redirect Graal log output to a file.]
[thread:27] scope: ForkJoinPool-4-worker-3
      [thread:27] scope: ForkJoinPool-4-worker-3.CreateBootImage.NativeBootImage.build.DebugCodeInfo
      Context: HostedMethod<JavaConstant.toString -> AnalysisMethod<jdk.vm.ci.meta.JavaConstant.toString -> HotSpotMethod<JavaConstant.toString(JavaConstant)>>>
      PrimaryRange JavaConstant.java jdk.vm.ci.meta.JavaConstant.toString
[native-image:29453]        image:   8,048.42 ms,  6.93 GB
[native-image:29453]        write:     900.56 ms,  6.93 GB
[native-image:29453]      [total]:  70,516.26 ms,  6.93 GB
```
Using `-H:Log=DebugCodeInfo,~DebugCodeInfo.Subranges -H:MethodFilter=JavaConstant.toString` causes only to output logging for methods that match `JavaConstant.toString`. See https://github.com/oracle/graal/blob/master/compiler/src/org.graalvm.compiler.debug/src/org/graalvm/compiler/debug/doc-files/MethodFilterHelp.txt for more info.
`DebugCodeInfo,~DebugCodeInfo.Subranges` says log scopes that what match `DebugCodeInfo` but exclude what matches `DebugCodeInfo.Subranges`. For all details check https://github.com/oracle/graal/blob/master/compiler/src/org.graalvm.compiler.debug/src/org/graalvm/compiler/debug/doc-files/DumpHelp.txt